src/github.com/matrix-org/rageshake/submit.go

```diff
@@ -245,19 +277,39 @@ func parseFormPart(part *multipart.Part, p *payload) error {
 	return nil
 }
 
-func (s *submitServer) saveReport(ctx context.Context, p payload) (*submitResponse, error) {
-	resp := submitResponse{}
+// saveFormPart saves a file upload to the report directory.
+//
+// Returns the leafname of the saved file.
+func saveFormPart(originalFileName string, reader io.Reader, reportDir string) (string, error) {
+	// figure out a safe name. Replace slashes, other path separators, and
+	// leading dots with underscores.
+	leafName := strings.Replace(originalFileName, "/", "_", -1)
+	leafName = strings.Replace(leafName, string(filepath.Separator), "_", -1)
+	if leafName == "" {
+		leafName = "unnamed-file"
+	} else if leafName[0] == '.' {
+		leafName = "_" + leafName[1:]
+	}
```

Comment:
Hmm, I'd be a lot happier if we URI encoded the originalFileName here, or used a random ID to store the file on disk. Allowing users to supply file names that contain either special shell characters or terminal escape characters or invalid utf8 could be saddening.